### PR TITLE
Add Whisper language detection

### PR DESCRIPTION
**Pull request by ae9is**
See #302 

Adds support for automatically detecting language to Whisper tasks.

The existing HuggingFace and Whisper implementations in Python were used as reference:
[Hugging Face Transformers](https://github.com/huggingface/transformers/blob/98e8062df3cf82b367e8a243963e4afb0d9d3407/src/transformers/models/whisper/generation_whisper.py#L1425)
[Original Whisper](https://github.com/openai/whisper/blob/90db0de1896c23cbfaf0c58bc2d30665f709f170/whisper/decoding.py#L19)

Also updates the existing Whisper test suites, including adding a string similarity check for actual model output (as opposed to just output length). Please note that the "new" development dependency for these tests, "fastest-levenshtein" is already used by "webpack-cli".

### CODE REVIEW

**Comment by xenova** [comment]
Thanks for the PR! This will certainly be a useful feature. Regarding the implementation, I think it can be greatly simplified as follows:
- Instead of using `.generate`, perform a single forward pass of the inputs
- Then, consider all logits which correspond to the language token ids
- Choose the language with the highest score

Currently, the implementation seems to perform a full generation step.

**Comment by ZhangPeng4242** [comment]
hey there, please approve this feature, it is a quite useful feature :)

File: src/models.js [hunk 1]
```diff
@@ -3119,10 +3124,48 @@ export class WhisperForConditionalGeneration extends WhisperPreTrainedModel {
...
+            ...options,
+            generation_config: {
+                ...generation_config,
+                good_words_ids,
+                num_beams: 1,
+                do_sample: false,
+            },
+            stopping_criteria,
+            decoder_input_ids,
+        });
```
Comment (xenova) [inline src/models.js:3158]: We should be able to replace this with a single forward pass (by called `this.forward(...)` instead of using a generation step.

File: src/models.js [hunk 2]
```diff
@@ -3119,10 +3124,48 @@ export class WhisperForConditionalGeneration extends WhisperPreTrainedModel {
...
 
     /**
+     * Detects language by running input through the model and checking for language tokens in the output.
      * 
-     * @param {WhisperGenerationConfig} generation_config 
+     * @param {import('./models/whisper/generation_whisper.js').WhisperGenerationFunctionParameters} options
+     * @returns {Promise<number[]>} A list of language token IDs detected.
+     */
+    async _detect_language(options) {
+        const inputs = options.inputs
```
Comment (edbrdi) [inline src/models.js:3133]: When testing this PR "inputs" was in my case not present, instead I had "input_features".

I noticed the type returns: "(Tensor of varying shape depending on the modality, optional): The sequence used as a prompt for the generation or as model inputs to the encoder. If null the method initializes it with bos_token_id and a batch size of 1. For decoder-only models inputs should be in the format of input_ids. For encoder-decoder models inputs can represent any of input_ids, input_values, input_features, or pixel_values."

By the way thanks for adding language detection, hope it will be merged soon :)

Comment (edbrdi) [reply src/models.js:3133]: I think it depends on the model used, you probably can reproduce it with https://huggingface.co/onnx-community/whisper-large-v3-turbo - I was able to fix it with `const inputs = options.inputs ?? options.input_features;` in _detect_language on my side.

**Comment by vklyukin** [comment]
Hello! I wanted to check in on this PR - would it be possible to get this approved and merged soon as it would significantly improve user experience with automatic language detection?

**Comment by ZhangPeng4242** [comment]
Any updates on this merge? thank you.

**Comment by ae9is** [comment]
In case it helps anyone else out there: https://github.com/ae9is/transformers.js/tree/develop-3.7.x

Replayed my commits on main, plus some minor fixes and changes for my use case.

**Comment by ae9is** [comment]
For anyone still following this issue, a month ago I updated my fork including Whisper language autodetection for version 4 here: https://github.com/ae9is/transformers.js/tree/develop-4.0.0. See particularly [#3e2adc](https://github.com/ae9is/transformers.js/commit/3e2adc52dc700d7afb820c7dcde29031fe78a37e)

If the maintainers are still interested, I can update this PR. Perhaps best to wait until after #1594 

I've also noticed that since my changes another PR has popped up for v4 (haven't reviewed this at all): #1541

**Comment by ae9is** [comment]
Going to close this one out of respect for everyone's time.

Offer to update for v4 still stands if the maintainers are interested, just reach out!
